src/main/scala/com/fulcrumgenomics/umi/ConsensusCallingIterator.scala

```diff
@@ -0,0 +1,90 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.umi
+
+import com.fulcrumgenomics.util.ProgressLogger
+import htsjdk.samtools.SAMRecord
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+/**
+  * An iterator that consumes from an incoming queue of SAMRecords and generates consensus
+  * read SAMRecords using the supplied caller.
+  *
+  * @param iterator an iterator of SAMRecords
+  * @param caller   the consensus caller to use to call consensus reads
+  * @param progress an optional progress logger to which to log progress in input reads
+  */
+class ConsensusCallingIterator(iterator: Iterator[SAMRecord],
+                               val caller: UmiConsensusCaller[_],
+                               val progress: Option[ProgressLogger] = None
+                              ) extends Iterator[SAMRecord] {
+
+  private val input = iterator.buffered
+  private val outputQueue: mutable.Queue[SAMRecord] = mutable.Queue[SAMRecord]()
+
+    /** True if there are more consensus reads, false otherwise. */
+  def hasNext(): Boolean = this.outputQueue.nonEmpty || (this.input.nonEmpty && advance())
+
+  /** Returns the next consensus read. */
+  def next(): SAMRecord = {
+    if (!this.hasNext()) throw new NoSuchElementException("Calling next() when hasNext() is false.")
+    this.outputQueue.dequeue()
+  }
+
+  /**
+    * Consumes the next group of records from the input iterator, based on molecule id
+    * and returns then as a Seq.
+    */
+  private def nextGroupOfRecords: Seq[SAMRecord] = {
+    if (iterator.isEmpty) Nil
+    else {
+      val idToMatch = this.caller.sourceMoleculeId(this.input.head)
+      val buffer = ListBuffer[SAMRecord]()
+      while (this.input.hasNext && this.caller.sourceMoleculeId(this.input.head) == idToMatch) {
+        buffer += this.input.next()
+      }
+
+      buffer
+    }
+  }
+
+  /** Consumes input records until one or more consensus reads can be created, or no more input records are available.
+    * Returns true if a consensus read was created and enqueued, false otherwise. */
+  @annotation.tailrec
+  private def advance(): Boolean = {
+    // get the records to create the consensus read
+    val inputs  = nextGroupOfRecords
+    val outputs = this.caller.consensusReadsFromSamRecords(inputs)
+    this.outputQueue ++= outputs
+
+    // Log progress on the _input_ reads and then return/recurse
+    for (p <- progress; r <- inputs) p.record(r)
+    if (outputs.nonEmpty) true
+    else if (this.input.hasNext) advance()
```

Comment:
Yup.  The compiler will yell if it is not.  It's still tail recursive because the current stack has no use once we go down the if - it's returning the value from `advance()` and not evaluating anything else.